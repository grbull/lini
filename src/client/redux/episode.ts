/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { api } from '../utils/api';

type EpisodeState = {
  isLoading: boolean;
  error?: string;
  data?: EpisodeDto;
};

const initialState: EpisodeState = {
  isLoading: false,
};

export const get = createAsyncThunk<EpisodeDto, number>('episode/get', (id) =>
  api.episode.get(id)
);

export const episode = createSlice({
  name: 'episode',
  initialState,
  reducers: {
    clear: (state) => {
      state.data = undefined;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.isLoading = false;
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.isLoading = false;
      state.error = error.message;
    });
  },
});

export const episodeActions = { ...episode.actions, get };
