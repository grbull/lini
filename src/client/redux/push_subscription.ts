/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { PushSubscriptionCreateDto } from '../../server/push_subscription/push_subscription.dto';
import { api } from '../utils/api';

type PushSubscriptionState = {
  status: 'loading' | 'idle' | 'error';
  error?: string;
};

const initialState: PushSubscriptionState = {
  status: 'idle',
};

export const create = createAsyncThunk<void, PushSubscriptionCreateDto>(
  'push_subscription/get',
  (createDto) => api.push.create(createDto)
);

export const pushSubscription = createSlice({
  name: 'push_subscription',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(create.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(create.fulfilled, (state) => {
      state.status = 'idle';
    });
    builder.addCase(create.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const pushSubscriptionActions = {
  ...pushSubscription.actions,
  create,
};
