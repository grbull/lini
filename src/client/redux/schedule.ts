/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import {
  EpisodeDto,
  EpisodeScheduleDto,
} from '../../server/episode/episode.dto';
import { api } from '../utils/api';

type ScheduleState = {
  status: 'init' | 'loading' | 'idle' | 'error';
  data: {
    future: EpisodeDto[];
    past: EpisodeDto[];
  };
  error?: string;
};

const initialState: ScheduleState = {
  status: 'init',
  data: {
    future: [],
    past: [],
  },
};

export const get = createAsyncThunk<EpisodeScheduleDto>('schedule/get', () =>
  api.episode.getSchedule()
);

export const schedule = createSlice({
  name: 'schedule',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const scheduleActions = {
  ...schedule.actions,
  get,
};
