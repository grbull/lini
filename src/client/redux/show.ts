/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { ShowWithEpisodesDto } from '../../server/show/show.dto';
import { api } from '../utils/api';

type ShowState = {
  status: 'init' | 'loading' | 'idle' | 'error';
  data?: ShowWithEpisodesDto;
  error?: string;
};

const initialState: ShowState = {
  status: 'init',
};

export const get = createAsyncThunk<ShowWithEpisodesDto, number>(
  'show/get',
  (id) => api.show.get(id)
);

export const show = createSlice({
  name: 'show',
  initialState,
  reducers: {
    reset: () => {
      return initialState;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.status = 'loading';
      state.data = undefined;
      state.error = undefined;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const showActions = {
  ...show.actions,
  get,
};
