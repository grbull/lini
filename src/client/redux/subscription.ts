/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import {
  SubscriptionCreateDto,
  SubscriptionDto,
  SubscriptionRemoveDto,
} from '../../server/subscription/subscription.dto';
import { api } from '../utils/api';
import { scheduleActions } from './schedule';

type SubscriptionState = {
  status: 'init' | 'loading' | 'idle' | 'error';
  data: SubscriptionDto[];
  error?: string;
};

const initialState: SubscriptionState = {
  status: 'init',
  data: [],
};

export const getAll = createAsyncThunk<SubscriptionDto[]>(
  'subscription/getAll',
  () => api.subscription.getAll()
);

export const create = createAsyncThunk<SubscriptionDto, SubscriptionCreateDto>(
  'subscription/create',
  async (createDto, thunkAPI) => {
    const subscription = await api.subscription.create(createDto);
    thunkAPI.dispatch(scheduleActions.get());
    return subscription;
  }
);

export const remove = createAsyncThunk<SubscriptionDto, SubscriptionRemoveDto>(
  'subscription/remove',
  async (removeDto, thunkAPI) => {
    const subscription = await api.subscription.remove(removeDto);
    thunkAPI.dispatch(scheduleActions.get());
    return subscription;
  }
);

export const subscription = createSlice({
  name: 'subscription',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(getAll.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(getAll.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data = payload;
    });
    builder.addCase(getAll.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });

    builder.addCase(create.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(create.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data.push(payload);
    });
    builder.addCase(create.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });

    builder.addCase(remove.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(remove.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data.splice(
        state.data.findIndex((s) => s.show.id === payload.show.id)
      );
    });
    builder.addCase(remove.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const subscriptionActions = {
  ...subscription.actions,
  getAll,
  create,
  remove,
};
