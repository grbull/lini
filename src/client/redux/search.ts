/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { ShowDto } from '../../server/show/show.dto';
import { api } from '../utils/api';

type SearchState = {
  isLoading: boolean;
  error?: string;
  data?: ShowDto[];
};

const initialState: SearchState = {
  isLoading: false,
};

export const query = createAsyncThunk<ShowDto[], string>(
  'search/query',
  (query) => api.show.search(query)
);

export const search = createSlice({
  name: 'search',
  initialState,
  reducers: {
    clear: (state) => {
      state.data = undefined;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(query.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(query.fulfilled, (state, { payload }) => {
      state.isLoading = false;
      state.data = payload;
    });
    builder.addCase(query.rejected, (state, { error }) => {
      state.isLoading = false;
      state.error = error.message;
    });
  },
});

export const searchActions = { ...search.actions, query };
