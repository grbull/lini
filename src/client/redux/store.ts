import {
  combineReducers,
  configureStore,
  getDefaultMiddleware,
} from '@reduxjs/toolkit';
import {
  FLUSH,
  PAUSE,
  PERSIST,
  persistReducer,
  persistStore,
  PURGE,
  REGISTER,
  REHYDRATE,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import { episode } from './episode';
import { pushSubscription } from './push_subscription';
import { schedule } from './schedule';
import { search } from './search';
import { show } from './show';
import { subscription } from './subscription';
import { user } from './user';

export const reducers = combineReducers({
  user: user.reducer,
  search: search.reducer,
  subscription: subscription.reducer,
  show: show.reducer,
  episode: episode.reducer,
  schedule: schedule.reducer,
  pushSubscription: pushSubscription.reducer,
});

const persistConfig = {
  key: 'root',
  storage,
  version: 2,
};

const persistedReducer = persistReducer(persistConfig, reducers);

export const store = configureStore({
  reducer: persistedReducer,
  // https://github.com/rt2zz/redux-persist/issues/988#issuecomment-552242978
  middleware: getDefaultMiddleware({
    serializableCheck: {
      ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
    },
  }),
  devTools: process.env.NODE_ENV === 'development',
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
