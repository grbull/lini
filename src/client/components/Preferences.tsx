import React, { ChangeEvent, ReactElement } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { UserUpdateDto } from '../../server/user/user.dto';
import { useNotifications } from '../hooks/useNotifications';
import { RootState } from '../redux/store';
import { userActions } from '../redux/user';
import { PreferencesItem } from './PreferencesItem';
import { Select } from './Select';
import { SettingsCardRow } from './SettingsCardRow';
import { Toggle } from './Toggle';

export function Preferences(): ReactElement {
  const dispatch = useDispatch();
  const { isEnabled, toggle } = useNotifications();
  const { profile } = useSelector((state: RootState) => state.user);

  function handleTheme(event: ChangeEvent<HTMLSelectElement>): void {
    const userUpdateDto: UserUpdateDto = {
      theme: event.target.value as 'auto' | 'light' | 'dark',
      notifications: profile?.notifications || false,
      dataSaving: profile?.dataSaving || false,
    };
    dispatch(userActions.update(userUpdateDto));
  }

  function handleDataSaving(event: ChangeEvent<HTMLInputElement>): void {
    const userUpdateDto: UserUpdateDto = {
      theme: profile?.theme || 'auto',
      notifications: profile?.notifications || false,
      dataSaving: event.target.checked,
    };
    dispatch(userActions.update(userUpdateDto));
  }

  return (
    <>
      <SettingsCardRow>
        <PreferencesItem description="Some description" title="Theme">
          <Select onChange={handleTheme} value={profile?.theme} />
        </PreferencesItem>
      </SettingsCardRow>
      <SettingsCardRow>
        <PreferencesItem
          description="When enabled, displays push notifications regarding soon to be aired episodes."
          title="Notifications"
        >
          <Toggle id="test" onChange={toggle} value={isEnabled} />
        </PreferencesItem>
      </SettingsCardRow>
      <SettingsCardRow>
        <PreferencesItem
          description="Reduces network data usage by using loading lower resolution images."
          title="Data Saving"
        >
          <Toggle
            id="test"
            onChange={handleDataSaving}
            value={profile?.dataSaving || false}
          />
        </PreferencesItem>
      </SettingsCardRow>
    </>
  );
}
