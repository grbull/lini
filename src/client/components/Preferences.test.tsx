/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom/extend-expect';

import { configureStore } from '@reduxjs/toolkit';
import {
  fireEvent,
  render,
  RenderResult,
  waitFor,
} from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';

import { reducers } from '../redux/store';
import { Preferences } from './Preferences';

function setup(): RenderResult & { themeSelect: HTMLSelectElement } {
  const store = configureStore({
    reducer: reducers,
    preloadedState: {
      user: {
        status: 'idle',
        data: {
          theme: 'auto',
          email: 'test@email.com',
          notifications: false,
          dateCreated: '',
          dateModified: '',
        },
      },
    },
  });

  // Mock service worker
  Object.defineProperties(navigator, {
    serviceWorker: {
      value: {
        ready: {
          pushManager: {
            getSubscription: () => null,
            subscribe: jest.fn(),
          },
        },
      },
      writable: true,
    },
  });

  Object.defineProperties(window, {
    Notification: {
      value: {
        permission: 'denied',
      },
      writable: true,
    },
  });

  // I would have liked to confirm dispatch was called but I don't know a
  // a good way to mock it yet.
  // const dispatch = jest.fn();
  // store.dispatch = jest.fn().mockImplementation(() => dispatch);
  // jest.mock('react-redux', () => ({
  //   useDispatch: jest.fn(() => dispatch),
  // }));

  const utils = render(
    <Provider store={store}>
      <Preferences />
    </Provider>
  );

  const themeSelect = utils.getByLabelText('theme-select') as HTMLSelectElement;

  return { themeSelect, ...utils };
}

describe('Preferences Component', () => {
  it('matches the snapshot', () => {
    const { asFragment } = setup();
    expect(asFragment()).toMatchSnapshot();
  });

  it('changes theme correctly', async () => {
    const { themeSelect } = setup();

    fireEvent.change(themeSelect, { target: { value: 'light' } });
    fireEvent.focusOut(themeSelect);

    await waitFor(() => expect(themeSelect.value).toBe('light'));
  });
});
