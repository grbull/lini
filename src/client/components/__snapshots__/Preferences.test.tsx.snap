// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preferences Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex justify-between items-center py-1"
  >
    <div
      class="flex flex-col"
    >
      <span
        class="font-medium"
      >
        Theme:
      </span>
      <span
        class="font-light"
      >
        When set to auto, the app will use your system setting's to determine the theme.
      </span>
    </div>
    <span>
      <select
        aria-label="theme-select"
        class="bg-gray-500 py-1 rounded"
        id="theme"
        name="theme"
      >
        <option
          selected=""
          value="auto"
        >
          Auto (system)
        </option>
        <option
          value="dark"
        >
          Dark
        </option>
        <option
          value="light"
        >
          Light
        </option>
      </select>
    </span>
  </div>
  <div
    class="flex justify-between items-center py-1"
  >
    <div
      class="flex flex-col"
    >
      <span
        class="font-medium"
      >
        Notifications:
      </span>
      <span
        class="font-light"
      >
        When enabled, displays push notifications regarding soon to be aired episodes.
      </span>
    </div>
    <span>
      <div
        class="relative inline-block w-10 mr-0 align-middle select-none transition duration-200 ease-in"
      >
        <input
          aria-label="toggle"
          class="toggle-checkbox absolute block w-6 h-6 rounded-full bg-white border-4 appearance-none cursor-pointer"
          id="test"
          name="test"
          type="checkbox"
        />
        <label
          class="toggle-label block overflow-hidden h-6 rounded-full bg-gray-300 cursor-pointer"
          for="test"
        />
      </div>
    </span>
  </div>
</DocumentFragment>
`;
